README.md
[![Build Status](https://travis-ci.org/karupanerura/toml.svg?branch=master)](https://travis-ci.org/karupanerura/toml)
# NAME

TOML - Parser for Tom's Obvious, Minimal Language.

# SYNOPSIS

    use TOML qw(from_toml to_toml);

    # Parsing toml
    my $toml = slurp("~/.foo.toml");
    my $data = from_toml($toml);

    # With error checking
    my ($data, $err) = from_toml($toml);
    unless ($data) {
        die "Error parsing toml: $err";
    }

    # Creating toml
    my $toml = to_toml($data); 

# DESCRIPTION

`TOML` implements a parser for Tom's Obvious, Minimal Language, as
defined at [https://github.com/mojombo/toml](https://github.com/mojombo/toml). `TOML` exports two
subroutines, `from_toml` and `to_toml`,

# FAQ

- How change how to de-serialize?

    You can change `$TOML::PARSER` for change how to de-serialize.

    example:

        use TOML;
        use TOML::Parser;

        local $TOML::PARSER = TOML::Parser->new(
            inflate_boolean => sub { $_[0] eq 'true' ? \1 : \0 },
        );

        my $data = TOML::from_toml('foo = true');

# FUNCTIONS

- from\_toml

    `from_toml` transforms a string containing toml to a perl data
    structure or vice versa. This data structure complies with the tests
    provided at [https://github.com/mojombo/toml/tree/master/tests](https://github.com/mojombo/toml/tree/master/tests).

    If called in list context, `from_toml` produces a (`hash`,
    `error_string`) tuple, where `error_string` is `undef` on
    non-errors. If there is an error, then `hash` will be undefined and
    `error_string` will contains (scant) details about said error.

- to\_toml

    `to_toml` transforms a perl data structure into toml-formatted
    string.

# SEE ALSO

[TOML::Parser](https://metacpan.org/pod/TOML::Parser)

# LICENSE

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; version 2.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02111-1301 USA

# AUTHOR

Darren Chamberlain <darren@cpan.org>

# CONTRIBUTORS

- Tokuhiro Matsuno <tokuhirom@cpan.org>
- Matthias Bethke <matthias@towiski.de>
- Sergey Romanov <complefor@rambler.ru>
- karupanerura <karupa@cpan.org>


Changes
Revision history for Perl extension TOML

{{$NEXT}}

0.97 2016-03-20T01:40:44Z

    - Don't escape forward slash. (thanks to Vickenty)

0.96 2015-02-18T14:37:31Z

    - Just import Exporter's "import" instead of inheriting from it. (thanks to dolmen)

0.95 2014-10-08T23:41:57Z

    - Required TOML::Parser 0.04 or later. (bug fixed in this version)

0.94 2014-05-22T13:28:16Z

    - Required TOML::Parser 0.03 or later. (many bugs are fixed in this version)

0.93 2014-05-18T14:33:11Z

    - Use TOML::Parser for parsing.
    - Now conforms to the specifications of CPAN::Changes::Spec.

0.92 2013-04-24T07:49:02-04:00

    - COPYING, MANIFEST, README: Added another contributor, and GPL.
    - Changelog: Add Changelog, per issues/5
    - TOML.pm: Added forgotten contributor (sorry!)
    - MANIFEST, MANIFEST.SKIP: Add MANIFEST.SKIP to MANIFEST, and add .travis.yml to it
    - .travis.yml: travis-ci.org config.
    - TOML.pm: Handle embedded \" in strings.
    - TOML.pm: Bump version number.
    - : Merge pull request -4 from tokuhirom/better-unescape Better unescape
    - TOML.pm, t/escape.t: Unescape characters correctly.
    - README: Fix formatting for Contributor section.
    - TOML.pm: Fix formatting for Contributor section.

0.91 2013-03-19T09:01:13-04:00

    - README: Updated README from docs.
    - t/reparse.t: Test roundtripping data -> toml -> data.
    - TOML.pm: Code cleanups and new documentation.
    - MANIFEST: Add new tests to manifest.
    - t/example.t: Not sure why the structure was named $yaml, that doesn't make sense.
    - : Merge pull request -3 from tokuhirom/master to_toml
    - t/to_toml.t: better test case
    - TOML.pm: Allow \z after number
    - TOML.pm, t/to_toml.t: TOML generator
    - README, TOML.pm: Update docs to reflect contributions.
    - TOML.pm, t/broken.t: Update test to emit line number and broken line, as recommended by Matthias Bethke
    - README: Update README with new docs.
    - MANIFEST, TOML.pm, t/broken.t: On error, from_toml returns (undef, $error) in list context, and undef in scalar context.
    - .gitignore: Ignore build stuff.

0.9 2013-02-25T09:18:27-08:00

    - : Merge pull request -1 from sergeyromanov/master Replace 'vars' pragma
    - Parser for Tom's Obvious, Minimal Language

lib/TOML.pm
package TOML;

# -------------------------------------------------------------------
# TOML - Parser for Tom's Obvious, Minimal Language.
#
# Copyright (C) 2013 Darren Chamberlain <darren@cpan.org>
# -------------------------------------------------------------------

use 5.008005;
use strict;
use warnings;
use Exporter 'import';

our ($VERSION, @EXPORT, @_NAMESPACE, $PARSER);

use B;
use Carp qw(croak);
use TOML::Parser 0.03;

$VERSION = "0.96";
@EXPORT = qw(from_toml to_toml);
$PARSER = TOML::Parser->new(inflate_boolean  => sub { $_[0] });

sub to_toml {
    my $stuff = shift;
    local @_NAMESPACE = ();
    _to_toml($stuff);
}

sub _to_toml {
    my ($stuff) = @_;

    if (ref $stuff eq 'HASH') {
        my $res = '';
        my @keys = sort keys %$stuff;
        for my $key (grep { ref $stuff->{$_} ne 'HASH' } @keys) {
            my $val = $stuff->{$key};
            $res .= "$key = " . _serialize($val) . "\n";
        }
        for my $key (grep { ref $stuff->{$_} eq 'HASH' } @keys) {
            my $val = $stuff->{$key};
            local @_NAMESPACE = (@_NAMESPACE, $key);
            $res .= sprintf("[%s]\n", join(".", @_NAMESPACE));
            $res .= _to_toml($val);
        }
        return $res;
    } else {
        croak("You cannot convert non-HashRef values to TOML");
    }
}

sub _serialize {
    my $value = shift;
    my $b_obj = B::svref_2object(\$value);
    my $flags = $b_obj->FLAGS;

    return $value
        if $flags & ( B::SVp_IOK | B::SVp_NOK ) and !( $flags & B::SVp_POK ); # SvTYPE is IV or NV?

    my $type = ref($value);
    if (!$type) {
        return string_to_json($value);
    } elsif ($type eq 'ARRAY') {
        return sprintf('[%s]', join(", ", map { _serialize($_) } @$value));
    } elsif ($type eq 'SCALAR') {
        if (defined $$value) {
            if ($$value eq '0') {
                return 'false';
            } elsif ($$value eq '1') {
                return 'true';
            } else {
                croak("cannot encode reference to scalar");
            }
        }
        croak("cannot encode reference to scalar");
    }
    croak("Bad type in to_toml: $type");
}

my %esc = (
    "\n" => '\n',
    "\r" => '\r',
    "\t" => '\t',
    "\f" => '\f',
    "\b" => '\b',
    "\"" => '\"',
    "\\" => '\\\\',
    "\'" => '\\\'',
);
sub string_to_json {
    my ($arg) = @_;

    $arg =~ s/([\x22\x5c\n\r\t\f\b])/$esc{$1}/g;
    $arg =~ s/([\x00-\x08\x0b\x0e-\x1f])/'\\u00' . unpack('H2', $1)/eg;

    return '"' . $arg . '"';
}

sub from_toml {
    my $string = shift;
    local $@;
    my $toml = eval { $PARSER->parse($string) };
    return wantarray ? ($toml, $@) : $toml;
}

1;

__END__

=encoding utf-8

=for stopwords versa

=head1 NAME

TOML - Parser for Tom's Obvious, Minimal Language.

=head1 SYNOPSIS

    use TOML qw(from_toml to_toml);

    # Parsing toml
    my $toml = slurp("~/.foo.toml");
    my $data = from_toml($toml);

    # With error checking
    my ($data, $err) = from_toml($toml);
    unless ($data) {
        die "Error parsing toml: $err";
    }

    # Creating toml
    my $toml = to_toml($data); 

=head1 DESCRIPTION

C<TOML> implements a parser for Tom's Obvious, Minimal Language, as
defined at L<https://github.com/mojombo/toml>. C<TOML> exports two
subroutines, C<from_toml> and C<to_toml>,

=head1 FAQ

=over 4

=item How change how to de-serialize?

You can change C<$TOML::PARSER> for change how to de-serialize.

example:

    use TOML;
    use TOML::Parser;

    local $TOML::PARSER = TOML::Parser->new(
        inflate_boolean => sub { $_[0] eq 'true' ? \1 : \0 },
    );

    my $data = TOML::from_toml('foo = true');

=back

=head1 FUNCTIONS

=over 4

=item from_toml

C<from_toml> transforms a string containing toml to a perl data
structure or vice versa. This data structure complies with the tests
provided at L<https://github.com/mojombo/toml/tree/master/tests>.

If called in list context, C<from_toml> produces a (C<hash>,
C<error_string>) tuple, where C<error_string> is C<undef> on
non-errors. If there is an error, then C<hash> will be undefined and
C<error_string> will contains (scant) details about said error.

=item to_toml

C<to_toml> transforms a perl data structure into toml-formatted
string.

=back

=head1 SEE ALSO

L<TOML::Parser>

=head1 LICENSE

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; version 2.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02111-1301 USA

=head1 AUTHOR

Darren Chamberlain <darren@cpan.org>

=head1 CONTRIBUTORS

=over 4

=item Tokuhiro Matsuno <tokuhirom@cpan.org>

=item Matthias Bethke <matthias@towiski.de>

=item Sergey Romanov <complefor@rambler.ru>

=item karupanerura <karupa@cpan.org>

=back
